### PR TITLE
[MCC-268574] Make ZipkinTracer independent from IOwinContext (and Owin itself)

### PR DESCRIPTION
**Pull request by ghost**
This is my take on making the core ZipkinTracer independent from the `IOwinContext` - by make it dependent on the _good ol'_ `System.Web.HttpContext` 😮.

The reason why we need this, because structure of the ZipkinTracer does not allow us to remove the dependency on some http context. By make it dependent on the `HttpContext` (and structuring the framework-dependent code - i.e. _Owin_, _WebApi_ - to separate packages) we can have a core tracer which is dependent on the .NET Framework provided context only. At the same time, the framework dependent code will be able to retrofit the incoming requests to the `HttpContext` world on the fly (unfortunately though we won't be able to support self hosting Owin scenarios anymore as we need IIS to build our `HttpContext`).

Please note that this version does not introduce any new functionality but it has breaking changes, therefore we have to version it as the next major (v4.0.0) version.

@mdsol/team-51 Please review and merge.
@jcarres-mdsol @cabbott FYI

### CODE REVIEW

**Comment by bvillanueva-mdsol** [comment]
@lschreck-mdsol I tried finding Readme changes. Do we need to add instructions for WebAPI nuget?

File: src/Medidata.ZipkinTracer.Core/TraceProvider.cs [hunk 1]
```diff
@@ -45,7 +45,7 @@ internal class TraceProvider : ITraceProvider
         /// </summary>
         /// <param name="config">ZipkinConfig instance</param>
         /// <param name="context">the IOwinContext</param>
-        internal TraceProvider(IZipkinConfig config, IOwinContext context = null)
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.Core/TraceProvider.cs:48]: TODO note above looks irrelevant now. I think we can erase that.

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs [hunk 2]
```diff
@@ -34,7 +34,7 @@ static SpanCollector GetInstance(Uri uri, uint maxProcessorBatchSize)
             return instance;
         }
 
-        public ZipkinClient(IZipkinConfig zipkinConfig, IOwinContext context, SpanCollector collector = null)
+        public ZipkinClient(IZipkinConfig zipkinConfig, HttpContextBase context, SpanCollector collector = null)
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs:37]: Just want to confirm the adhoc if it covered these scenarios:
1. For Owin application, is zipkin properly logging for both incoming and outgoing?
2. For Web API application, is zipkin properly logging for both incoming and outgoing?

File: src/Medidata.ZipkinTracer.Owin/ZipkinTracerOwinExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Web;
+using Microsoft.Owin;
+
+namespace Medidata.ZipkinTracer.Owin
+{
+    internal static class ZipkinTracerOwinExtensions
+    {
+        public static readonly string HttpContextBaseKey = "System.Web.HttpContextBase";
```
Comment (h-kurniawan) [inline src/Medidata.ZipkinTracer.Owin/ZipkinTracerOwinExtensions.cs]: No magic string is better?
`public static readonly string HttpContextBaseKey = typeof(HttpContextBase).FullName;`

File: src/Medidata.ZipkinTracer.Owin/ZipkinTracerOwinExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+        public static HttpContextBase ToHttpContext(this IOwinContext owinContext) =>
+            owinContext.Environment.ContainsKey(HttpContextBaseKey) ?
+            owinContext.Environment[HttpContextBaseKey] as HttpContextBase :
+            throw new NotSupportedException("Self hosted and non-System.Web-based scenarios are not supported.");
```
Comment (h-kurniawan) [inline src/Medidata.ZipkinTracer.Owin/ZipkinTracerOwinExtensions.cs]: Actually the whole property can be simplified like
```cs
public static HttpContextBase ToHttpContext(this IOwinContext owinContext) =>
    owinContext.Get<HttpContextBase>(typeof(HttpContextBase).FullName) ??
    throw new NotSupportedException("Self hosted and non-System.Web-based scenarios are not supported.");
```
`IOwinContext.Get<T>` already does the checking of the key for you.
https://github.com/aspnet/AspNetKatana/blob/dev/src/Microsoft.Owin/OwinContext.cs#L88

File: src/Medidata.ZipkinTracer.Owin/ZipkinMiddleware.cs [hunk 5]
```diff
@@ -1,27 +1,32 @@
...
 using Microsoft.Owin;
 using Owin;
 
-namespace Medidata.ZipkinTracer.Core.Middlewares
+namespace Medidata.ZipkinTracer.Owin
 {
     public class ZipkinMiddleware : OwinMiddleware
     {
         private readonly IZipkinConfig _config;
+        private readonly SpanCollector _collector;
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.Owin/ZipkinMiddleware.cs]: Looks like _collector is always null here. ZipkinClient will take care the singleton instance of SpanCollector

Comment (bvillanueva-mdsol) [reply src/Medidata.ZipkinTracer.Owin/ZipkinMiddleware.cs]: Ow, did not see UseZipkin method is accepting SpanCollector. Is this giving the user of this library to maintain spanCollector?

File: src/Medidata.ZipkinTracer.WebApi/HttpContextContainer.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.WebApi/HttpContextContainer.cs]: minor comment: unused usings

File: src/Medidata.ZipkinTracer.WebApi/ZipkinTraceHandler.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (InnerHandler == null)
+                InnerHandler = new HttpClientHandler();
+
+            var context = HttpContextContainer.Current;
+
+            if (config.Bypass?.Invoke(context.Request) ?? false)
+                return await base.SendAsync(request, cancellationToken)
+                .ConfigureAwait(continueOnCapturedContext: false);
```
Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.WebApi/ZipkinTraceHandler.cs]: minor comment: more indention for '.ConfigureAwait(continueOnCapturedContext: false);'

**Review by bvillanueva-mdsol — CHANGES_REQUESTED** [review]
left some comments. thanks

File: tests/Medidata.ZipkinTracer.Tests/Handlers/HttpContextInjectorMiddleware.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web;
+using Medidata.ZipkinTracer.Core;
+using Medidata.ZipkinTracer.Owin;
+using Microsoft.Owin;
+using Owin;
+using Rhino.Mocks;
```
Comment (h-kurniawan) [inline tests/Medidata.ZipkinTracer.Tests/Handlers/HttpContextInjectorMiddleware.cs:13]: Not sure whether this is out of scope but, I think we need to move away from using Rhino Mocks whenever possible since it is no longer actively developed since 2011 according to author himself.
[The state of Rhino Mocks](https://ayende.com/blog/161826/the-state-of-rhino-mocks)

Comment (bvillanueva-mdsol) [reply tests/Medidata.ZipkinTracer.Tests/Handlers/HttpContextInjectorMiddleware.cs:13]: I agree, but might not be necessary for this PR. maybe on the big refactor PR 👍

File: src/Medidata.ZipkinTracer.Owin/ZipkinMiddleware.cs (hunk 5 above)
Comment (ghost) [reply src/Medidata.ZipkinTracer.Owin/ZipkinMiddleware.cs]: The point is to have this on the constructor is purely for the unit testing. Actually I think I can make the middleware internal and just have an internal `UseZipkin()` with the `SpanCollector`.

Edit: done - 3e26f94

File: tests/Medidata.ZipkinTracer.Tests/Handlers/HttpContextInjectorMiddleware.cs (hunk 8 above)
Comment (ghost) [reply tests/Medidata.ZipkinTracer.Tests/Handlers/HttpContextInjectorMiddleware.cs:13]: I would like to move away from the mocking frameworks altogether.

File: src/Medidata.ZipkinTracer.Core/TraceProvider.cs (hunk 1 above)
Comment (ghost) [reply src/Medidata.ZipkinTracer.Core/TraceProvider.cs:48]: Removed, 25961f7

File: src/Medidata.ZipkinTracer.Owin/ZipkinTracerOwinExtensions.cs (hunk 4 above)
Comment (ghost) [reply src/Medidata.ZipkinTracer.Owin/ZipkinTracerOwinExtensions.cs]: Simplified, thanks! 25961f7

File: src/Medidata.ZipkinTracer.WebApi/HttpContextContainer.cs (hunk 6 above)
Comment (ghost) [reply src/Medidata.ZipkinTracer.WebApi/HttpContextContainer.cs]: Removed, 25961f7 thanks!

File: src/Medidata.ZipkinTracer.WebApi/ZipkinTraceHandler.cs (hunk 7 above)
Comment (ghost) [reply src/Medidata.ZipkinTracer.WebApi/ZipkinTraceHandler.cs]: Thanks, 25961f7 indented!

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs (hunk 2 above)
Comment (ghost) [reply src/Medidata.ZipkinTracer.Core/ZipkinClient.cs:37]: Yes, I adhoced with simultaneous incoming/outgoing requests, and with Owin and WebAPI as well.

Now one thing which I should note, that when you use the `ZipkinMessageHandler` for the ougoing requests, you have to provide the `HttpContext` yourself (more specifically `HttpContextBase`) because the handler won't do it for you - this is really easy in Owin and WebAPI, you just need to wrap the `HttpContext.Current` object: `new HttpContextWrapper(HttpContext.Current)` - as for command line applications, this is not trivial but it is not impossible either to have a working HttpContext (it is just very inconvenient).

File: src/Medidata.ZipkinTracer.Owin/ZipkinTracerOwinExtensions.cs (hunk 3 above)
Comment (ghost) [reply src/Medidata.ZipkinTracer.Owin/ZipkinTracerOwinExtensions.cs]: Done, 25961f7

**Review by h-kurniawan — APPROVED** [review]

File: src/Medidata.ZipkinTracer.WebApi/ZipkinTraceHandler.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web;
+using Medidata.ZipkinTracer.Core;
+
+namespace Medidata.ZipkinTracer.WebApi
+{
+    public class ZipkinTraceHandler: DelegatingHandler
+    {
+		private readonly IZipkinConfig config;
```
Comment (kenyamat) [inline src/Medidata.ZipkinTracer.WebApi/ZipkinTraceHandler.cs]: please fix the indent

Comment (ghost) [reply src/Medidata.ZipkinTracer.WebApi/ZipkinTraceHandler.cs]: Done - 9344588

**Comment by ghost** [comment]
Closing the PR as we won't make this change here.
